godot-core/src/builtin/vector_macros.rs

```diff
@@ -0,0 +1,221 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+#![macro_use]
+
+/// Implements a single unary operator for a vector type. Only used for `Neg` at the moment.
+macro_rules! impl_vector_unary_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Neg`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `neg`.
+        $func:ident$(,)?
+    ) => {
+        // -vector
+        impl std::ops::$operator for $vector {
+            type Output = Self;
+            fn $func(mut self) -> Self::Output {
+                $(
+                    self.$components = self.$components.$func();
+                )*
+                self
+            }
+        }
+    }
+}
+
+/// Implements a single infix binary operator for a vector type, between two vectors and between a
+/// vector and a scalar in both directions.
+macro_rules! impl_vector_binary_operator {
+    (
+        // Name of the vector type.
+        $vector:ty,
+        // Type of each individual component, for example `i32`.
+        $scalar:ty,
+        // Names of the components, with parentheses, for example `(x, y)`.
+        ($($components:ident),*),
+        // Name of the operator trait, for example `Add`.
+        $operator:ident,
+        // Name of the function on the operator trait, for example `add`.
+        $func:ident$(,)?
```

Comment:
Minor thing: I used the convention that types are `PascalCase` in declarative macros, which helped identify them in the macro body.

It's enough if you do it for anything `:ty`, i.e. `$Vector` and `$Scalar`.
I wouldn't do it for `:ident`, as that may be concatenated.

Reply:
I like that!